### PR TITLE
#128 able-waterbuck

### PR DESCRIPTION
**Pull request by AbrahamFergie**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,25 @@
+#Team able-waterBuffalo -
+Lisa Bronwyn, Sal Soloman, Abraham Ferguson
+
+#Goals:
+- We will Write Test cases for each data structure and develop code to pass said tests.  All created through the javascript language.
+- Because each group member has done this project before we will be doing deeper dives on the last 5 data structures in order to further our understanding.
+
+  X- Stack
```
Comment (jrobcodes) [inline README.md:8]: This could be beautified in md using 
```md
[X] Stack
```

File: spec/directedGraph.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  })
+
+  context('addVertex()', () => {
+    it('adds a vertex to the graph.', () => {
+      const dg = new DirectedGraph()
+      for(let i = 0; i < 6; i++ ){
+        dg.addVertex(i)
+      }
+      dg.makeGraph()
+      console.log()
```
Comment (jrobcodes) [inline spec/directedGraph.js:19]: Remove `console.log` prior to submission.

File: spec/directedGraph.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    expect(DirectedGraph).to.be.a('function')
+  })
+
+  context('addVertex()', () => {
+    it('adds a vertex to the graph.', () => {
+      const dg = new DirectedGraph()
+      for(let i = 0; i < 6; i++ ){
+        dg.addVertex(i)
+      }
+      dg.makeGraph()
```
Comment (jrobcodes) [inline spec/directedGraph.js:18]: These two lines stood out to me (prior to looking at the implementation of the graph):
```js
dg.makeGraph()
dg.graphSize()
```
This implies that some side effect is required in order to cause the graph to function as expected...  If no such side effects exist, remove extraneous code.  I would also note that these functions do not appear in the specification for graph, as defined in the spec.

File: spec/doublyLinkedList.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      list.insert('another-element')
+
+      expect( list.find('another-element') ).to.eql( list.getTailNode() )
+    })
+  })
+  context('insertFirst()', () => {
+    it('Inserts a node (with the provided data) to the head of the list', () => {
+      const list = new DoublyLinkedList()
+
+      list.insert('element')
```
Comment (jrobcodes) [inline spec/doublyLinkedList.js:27]: Consider only inserting enough elements to test for the actual functionality - additional code may decrease readability for maintainers.

File: spec/doublyLinkedList.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    it('Inserts a node (with data "apples") before the first node containing "bananas"', () => {
+      const list = new DoublyLinkedList()
+
+      list.insert('element')
+      list.insert('other-element')
+      list.insert('other-elementals')
+      list.insert('elementals')
+      list.insert('bananas')
+
+      // expect( list.size() ).to.eql( 5 )
```
Comment (jrobcodes) [inline spec/doublyLinkedList.js:47]: Remove comments.

File: spec/doublyLinkedList.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      expect( list.insertBefore('bananas', 'apples' ) ).to.eql(
```
Comment (jrobcodes) [inline spec/doublyLinkedList.js:48]: This is another red flag to me, prior to looking at the implementation - why is `insertBefore` returning an object literal (that appears to reveal the underlying data representation of the list)?  We should be using the list API in order to effect our tests.

File: spec/hashTable.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  })
+  context('put()', () => {
+    it('adds a key-value pair to the hash table.', () => {
+      const ht = new HashTable()
+      ht.put('name', 'frankenstein')
+      expect(ht.size()).to.equal(1)
+      ht.put('19.0', 'stein')
+      expect(ht.size()).to.equal(2)
+      ht.get('19.0')
+      // expect(ht.get('19.0')).to.eql('stein')
```
Comment (jrobcodes) [inline spec/hashTable.js:19]: Is this broken? 😨  Curious why a test was commented out...

**Comment by lizzkats** [comment]
Great job! Love the uploaded notes :)

File: spec/hashTable.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+describe('hashTable', () => {
+  it('exists', () => {
+    expect(HashTable).to.be.a('function')
+  })
+  context('put()', () => {
+    it('adds a key-value pair to the hash table.', () => {
+      const ht = new HashTable()
+      ht.put('name', 'frankenstein')
+      expect(ht.size()).to.equal(1)
```
Comment (jrobcodes) [inline spec/hashTable.js:15]: This just tests that the size of the hash table changes on `put` - I would test this separately.  We would also need a test to verify that the key value pair was stored in the table (perhaps using the `get` function).

File: spec/hashTable.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+      ht.put('19.0', 'stein')
+      expect(ht.size()).to.equal(2)
+      ht.get('19.0')
+      // expect(ht.get('19.0')).to.eql('stein')
+    })
+  })
+  context('contains()', () => {
+    it('returns true if the hash table contains the key.', () => {
+      const ht = new HashTable()
+      expect(ht.contains('name')).to.be.false
```
Comment (jrobcodes) [inline spec/hashTable.js:25]: I recommend making these are two separate tests - one that tests that non-existent keys return false, the other to test for keys in hash.  This isolates each of the cases that can come from calling `contains` into their own test.

File: spec/hashTable.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    })
+  })
+  context('remove()', () => {
+    it('removes a key-value pair by key.', () => {
+      const ht = new HashTable()
+      ht.put('name', 'frankenstein')
+      ht.put('19.0', 'stein')
+      ht.put('wsrd', 'the weekend')
+      ht.remove('19.0')
+      console.log('just removed 19.0', ht.dataStore);
```
Comment (jrobcodes) [inline spec/hashTable.js:38]: Remove `cosole.log`.

File: spec/hashTable.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+      expect(ht.contains('19.0')).to.be.false
+    })
+  })
+  context('iterate()', () => {
+    it('takes a callback function and passes it each key and value in sequence.', () => {
+      const ht = new HashTable()
+      ht.put('name', 'frankenstein')
+      ht.put('19.0', 'stein')
+      ht.put('wsrd', 'the weekend')
+      let arr = ht.iterate('19.0')
```
Comment (jrobcodes) [inline spec/hashTable.js:49]: `iterate` does not appear to be taking a callback function here?  As such, I'm not sure I understand the intent of this test...

File: spec/helpers.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
+import linkedList from '../src/linkedList'
```
Comment (jrobcodes) [inline spec/helpers.js:1]: Type names should be upper camel case (`LinkedList`).  I initially interpreted this as a variable identifier.

File: spec/linkedList.js [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+
+chai.use(chaiChange)
+
+describe('linkedList', () => {
+  it('exists', () => {
+    expect(linkedList).to.be.a('function')
+  })
+
+  context('insert()', () => {
+    it('creates a new node and places in the linked list as the new tail node', () => {
```
Comment (jrobcodes) [inline spec/linkedList.js:13]: I recommend more concise test descriptions:
```js
it( 'inserts a new node as the tail' )
```

File: spec/linkedList.js [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+      const myLinkedList = new linkedList()
+      myLinkedList.insert( 'water' )
+      myLinkedList.insert( 'balloons' )
+
+      console.log(myLinkedList._length);
```
Comment (jrobcodes) [inline spec/linkedList.js:18]: Remove `console.log`

File: spec/linkedList.js [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+      expect(myLinkedList.head).to.eql({ data: 'water', next: { data: 'balloons', next: null}})
```
Comment (jrobcodes) [inline spec/linkedList.js:19]: These tests rely on the internal representation of the `LinkedList` - with this implementation, you will be unable to change the underlying data representation without breaking any client of this class.  Internal data should only be accessible via accessor functions.

File: spec/linkedList.js [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+  })
+  context('getHeadNode()' , () => {
+    it('returns the first node on the linked list', () => {
+      const myLinkedList = new linkedList()
+      expect(myLinkedList.getHeadNode()).to.eql(null)
+      myLinkedList.insert( 'water' )
+      myLinkedList.insert( 'balloons' )
+      myLinkedList.insert( 'lion' )
+
+      expect(myLinkedList.getHeadNode().data).to.eql( 'water' )
```
Comment (jrobcodes) [inline spec/linkedList.js:31]: I recommend the use of accessors for accessing data internal to the class, in this case:
```js
class Node {
  constructor( data, next=null ) {
    this.data = data
    this.next = next
  }

  getData() {
    return this.data
  }
}
```

File: spec/linkedList.js [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+    })
+  })
+  context('getTailNode()', () => {
+    it('returns the last node on the linked list', () => {
+      const myLinkedList = new linkedList()
+      expect(myLinkedList.getTailNode()).to.eql(null)
```
Comment (jrobcodes) [inline spec/linkedList.js:37]: This is a separate test - the expectation that an empty list returns `null` for a call to `getTailNode`.

File: spec/linkedList.js [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+  context('insertFirst()', () => {
+    it('Inserts a node (with the provided data) to the head of the list', () => {
+      const myLinkedList = new linkedList()
+      myLinkedList.insert( 'muffin' )
+      myLinkedList.insert( 'steel' )
+      expect(myLinkedList._length).to.eql(2)
+
+      myLinkedList.insertFirst( 'bear' )
+
+      expect(myLinkedList.head.data).to.eql('bear')
```
Comment (jrobcodes) [inline spec/linkedList.js:78]: Should not be using private data in test.

File: spec/linkedList.js [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+  })
+  context('insertBefore()', () => {
+    it('Inserts a node (with data "glasses") before the first node containing "atom"', () => {
+      const myLinkedList = new linkedList()
+      myLinkedList.insert( 'muffin' )
+      myLinkedList.insert( 'steel' )
+      myLinkedList.insert( 'atom' )
+      myLinkedList.insert('lions')
+      expect(myLinkedList._length).to.eql(4)
+      expect(myLinkedList.insertBefore('atom', 'glasses').next.data).to.eql('atom')
```
Comment (jrobcodes) [inline spec/linkedList.js:90]: Should not be using private data in test.  As well, this is a violation of the [Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) - the test needs to understand, and is dependent on the stability of, the internal structure of the node.

File: spec/linkedList.js [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+  context('remove()', () => {
+    it('Removes the tail node from the list', () => {
+      const myLinkedList = new linkedList()
+      myLinkedList.insert( 'muffin' )
+      myLinkedList.insert( 'steel' )
+      myLinkedList.insert( 'atom' )
+      myLinkedList.insert('lions')
+      expect(myLinkedList._length).to.eql(4)
+      myLinkedList.remove('lions')
+      console.log('---------', myLinkedList);
```
Comment (jrobcodes) [inline spec/linkedList.js:115]: Remove `console.log`

File: spec/linkedList.js [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+  context('removesFirst()', () => {
+    it('Removes the head node from the list', () => {
+      const myLinkedList = new linkedList()
+      myLinkedList.insert( 'muffin' )
+      myLinkedList.insert( 'steel' )
+      myLinkedList.insert( 'atom' )
+      myLinkedList.insert('lions')
+      expect(myLinkedList._length).to.eql(4)
+      myLinkedList.removesFirst()
+      console.log('---------', myLinkedList);
```
Comment (jrobcodes) [inline spec/linkedList.js:128]: Remove `console.log`

File: spec/linkedList.js [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+    })
+  })
+  context('clears()', () => {
+    it('Clears the list of all nodes/data', () => {
+      const myLinkedList = new linkedList()
+      myLinkedList.insert( 'muffin' )
+      myLinkedList.insert( 'steel' )
+      myLinkedList.insert( 'atom' )
+      myLinkedList.insert('lions')
+      console.log('!!!!=========', myLinkedList)
```
Comment (jrobcodes) [inline spec/linkedList.js:151]: Remove `console.log`

File: spec/linkedList.js [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+
+      myLinkedList.clears()
+      console.log('=========', myLinkedList)
```
Comment (jrobcodes) [inline spec/linkedList.js:154]: Remove `console.log`

File: spec/linkedList.js [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+      const myLinkedList = new linkedList()
+      expect(myLinkedList.isEmpty()).to.be.true
+      myLinkedList.insert( 'muffin' )
+      myLinkedList.insert( 'steel' )
+      myLinkedList.insert( 'atom' )
+      myLinkedList.insert('lions')
+      expect(myLinkedList.isEmpty()).to.be.false
+    })
+  })
+  context('clears()', () => {
```
Comment (jrobcodes) [inline spec/linkedList.js:144]: I think the spec requires this method to be named `clear`?

File: spec/pQueue.js [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+      thePQueue.enqueue({ priority: 400, data: 'fooDoo' })
+
+      expect( thePQueue.dequeue() ).to.eql( { priority: 12, data: 'foo' } )
+    })
+  })
+
+  context('isEmpty()', () => {
+    it('returns true if the queue is empty or false if not.', () => {
+      const thePQueue = new pQueue()
+      expect( thePQueue.isEmpty() ).to.eql( true )
```
Comment (jrobcodes) [inline spec/pQueue.js:68]: These are two separate tests.

File: spec/queue.js [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+
+chai.use(chaiChange)
+
+describe('Queue', () => {
+  it('exists', () => {
+    expect(Queue).to.be.a('function')
+  })
+
+  context('enqueue()', () => {
+    it('adds an element (the string "foo") to the back of the queue.', () => {
```
Comment (jrobcodes) [inline spec/queue.js:13]: The test description should specify the behavior, not the implementation of the test.  I recommend removing `(the string "foo")` from the description.

File: spec/queue.js [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+      const aQueue = new Queue()
+      aQueue.enqueue( 'foo' )
+      aQueue.enqueue( 'fooBar' )
+      aQueue.enqueue( 'fooBoo' )
+
+      expect( aQueue.sequence ).to.eql( [ 'foo','fooBar', 'fooBoo' ] )
```
Comment (jrobcodes) [inline spec/queue.js:19]: What is `.sequence`?  This tests the underlying representation of data in the `Queue`, not necessarily the functionality of the `enqueue` method.

File: spec/set.js [hunk 28]
```diff
@@ -0,0 +1,209 @@
...
+describe('Set', () => {
+  it('exists', () => {
+    expect(Set).to.be.a('function')
+  })
+
+  context('add()', () => {
+    it('adds an element to the set.', () => {
+      const aSet = new Set([])
+
+      expect( () => aSet.add( 'foo' ) )
```
Comment (jrobcodes) [inline spec/set.js:16]: This only tests the size change, not whether or not the element is in the `Set`.  You also need a test for non-unique elements (the size should not change).

File: spec/set.js [hunk 29]
```diff
@@ -0,0 +1,209 @@
...
+        .to.alter( () => aSet.size(), { from: 0, to: 1 } )
+
+    })
+  })
+
+  context('isEmptyTrue()', () => {
```
Comment (jrobcodes) [inline spec/set.js:22]: This is confusing - it looks like a function?  The `context` should describe the context in which the test is occurring.  In this case:
```js
context( 'when the set is empty', () => { // etc.
```

File: spec/set.js [hunk 30]
```diff
@@ -0,0 +1,209 @@
...
+    it('returns true if the queue is empty or false if not.', () => {
+      const aSet = new Set([])
+      expect( aSet.isEmpty() ).to.eql( true )
+      expect( () => aSet.isEmpty() ).to.alter( () => aSet.size(), { from: 0, to: 0 } )
+    })
+  })
+
+  context('isEmptyFalse()', () => {
```
Comment (jrobcodes) [inline spec/set.js:30]: See above.  As a programmer, seeing and identifier with parentheses after makes this look like a function.

File: spec/set.js [hunk 31]
```diff
@@ -0,0 +1,209 @@
...
+    it('returns true if the queue is empty or false if not.', () => {
+      const aSet = new Set([])
+
+      expect( aSet.size() ).to.eql( 0 )
```
Comment (jrobcodes) [inline spec/set.js:34]: Multiple tests in one.

File: spec/stack.js [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+
+describe('Stack', () => {
+  'use strict'
+
+  it('exists', () => {
+    expect(Stack).to.be.a('function')
+  })
+
+  context('push()', () => {
+    it('pushes an element to the top of the stack.', () => {
```
Comment (jrobcodes) [inline spec/stack.js:15]: Missing a test for inclusion after element is pushed.

File: spec/stack.js [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+    })
+  })
+
+  context('pop()', () => {
+    it('pops an element off the top of the stack.', () => {
+      const myStack = new Stack()
+
+      myStack.push( 'fooBar' )
+      myStack.push( 'foo' )
+      expect( myStack.pop() ).to.eql( 'foo' )
```
Comment (jrobcodes) [inline spec/stack.js:29]: Missing a test for change of size on `pop`

File: spec/stack.js [hunk 34]
```diff
@@ -0,0 +1,62 @@
...
+      myStack.push( 'fooBar' )
+      expect( myStack.length() ).to.eql( 2 )
+    })
+  })
+
+  context('isEmpty()', () => {
+    it('returns a boolean stating whether there are elements in the array', () => {
+      const myStack = new Stack()
+
+      expect( myStack.isEmpty() ).to.eql( true )
```
Comment (jrobcodes) [inline spec/stack.js:47]: These should be two separate tests.

File: spec/stack.js [hunk 35]
```diff
@@ -0,0 +1,62 @@
...
+      myStack.push( 'foo' )
+      expect( myStack.isEmpty() ).to.eql( false )
+    })
+  })
+
+  context('peek()', () => {
+    it('returns top element in the stack, otherwise return null', () => {
```
Comment (jrobcodes) [inline spec/stack.js:54]: Even the description of the test states that this is two separate tests - I recommend splitting this into two separate tests...

File: src/directedGraph.js [hunk 36]
```diff
@@ -0,0 +1,92 @@
+export default class DirectedGraph {
```
Comment (jrobcodes) [inline src/directedGraph.js:1]: Skipping review since not completed and running out of time.

File: src/doublyLinkedList.js [hunk 37]
```diff
@@ -0,0 +1,103 @@
...
+  constructor() {
+    this.head = null
+    this._length = 0
+  }
+  getHeadNode() { return this.head }
+
+  getTailNode() {
+    let currentNode = this.head
+    console.log(currentNode)
+    // currentNode.next = this.next
```
Comment (jrobcodes) [inline src/doublyLinkedList.js:22]: Remove comments

File: src/doublyLinkedList.js [hunk 38]
```diff
@@ -0,0 +1,103 @@
...
+  }
+  data() { return this.data }
+  next() { return this.next }
+  previous() { return this.prev }
+}
+
+export default class DoublyLinkedList{
+  constructor() {
+    this.head = null
+    this._length = 0
```
Comment (jrobcodes) [inline src/doublyLinkedList.js:15]: This is a convention from really old C programs - to use the `_` for "private" variables.  I'm not a fan (especially in javascript where it could easily be confused for the lodash library), but if you're going to use it, I recommend using it consistently for private data (i.e. `head` should be `_head`).

File: src/doublyLinkedList.js [hunk 39]
```diff
@@ -0,0 +1,103 @@
...
+  }
+  getHeadNode() { return this.head }
+
+  getTailNode() {
+    let currentNode = this.head
+    console.log(currentNode)
```
Comment (jrobcodes) [inline src/doublyLinkedList.js:21]: Remove `console.log`

File: src/doublyLinkedList.js [hunk 40]
```diff
@@ -0,0 +1,103 @@
...
+
+    return false
+  }
+
+  find( data ) {
+    if(this.head){
+      let currentNode = this.head
+
+      while( currentNode.next ) {
+        // console.log(currentNode)
```
Comment (jrobcodes) [inline src/doublyLinkedList.js:50]: Remove comments

File: src/doublyLinkedList.js [hunk 41]
```diff
@@ -0,0 +1,103 @@
...
+        if( currentNode.data === data ){
+          return currentNode
+        } else {
+          currentNode = currentNode.next
+        }
+      }
+      return -1
```
Comment (jrobcodes) [inline src/doublyLinkedList.js:57]: Since this is repeated, I would refactor this to the end of the function, outside the `if/else` block.

File: src/doublyLinkedList.js [hunk 42]
```diff
@@ -0,0 +1,103 @@
...
+      this._length++
+      node.prev = null
+      return node
+    }
+    while( currentNode.next ) {
+      currentNode = currentNode.next
+    }
+    currentNode.next = node
+    node.prev = currentNode
+    // console.log('data',node);
```
Comment (jrobcodes) [inline src/doublyLinkedList.js:78]: Remove comments

File: src/doublyLinkedList.js [hunk 43]
```diff
@@ -0,0 +1,103 @@
...
+    this._length++
+
+    return node
+  }
+  insertBefore() {
```
Comment (jrobcodes) [inline src/doublyLinkedList.js:83]: :/ Double Linked List is marked as complete in the README...

File: src/hashTable.js [hunk 44]
```diff
@@ -0,0 +1,135 @@
...
+  put(key, value) {
+    let hashed = this.hashFunction(key)
+    ++this._length
+    if(!this.dataStore[hashed]){
+      this.save(hashed, key, value)
+    }else{
+      this.chain(hashed, key, value)
+    }
+  }
+  size() { return this._length }
```
Comment (jrobcodes) [inline src/hashTable.js:50]: I recommend avoiding the use of inlined functions to increase readability.

File: src/hashTable.js [hunk 45]
```diff
@@ -0,0 +1,135 @@
...
+        pair = this.dataStore[hashed][i]
+        if(pair.key === key)
+          return pair.value
+      }
+    }
+  }
+
+  contains(key) {
+    var hash = this.hashFunction(key)
+    //go through each of the nodes in the linked list at key location
```
Comment (jrobcodes) [inline src/hashTable.js:66]: Remove comments

File: src/hashTable.js [hunk 46]
```diff
@@ -0,0 +1,135 @@
...
+          if(currentNode.key === key){
+            return true
+          }else {
+            currentNode = currentNode.next
+          }
+        }
+      }
+      if(currentNode.key === key){
+        return true
+      }else { return false }
```
Comment (jrobcodes) [inline src/hashTable.js:82]: Very difficult to read.  Please expand code blocks.

File: src/linkedList.js [hunk 47]
```diff
@@ -0,0 +1,112 @@
...
+
+    while(currentNode.next){
+      currentNode = currentNode.next
+    }
+      ++this._length
+      currentNode.next = node
+    return this.tail = node
+  }
+  
+  size() { return this._length }
```
Comment (jrobcodes) [inline src/linkedList.js:32]: Expand code blocks in this class.

File: src/linkedList.js [hunk 48]
```diff
@@ -0,0 +1,112 @@
...
+    if(this.head === null){
+      ++this._length
+      return this.head = this.tail = node
+    }
+    let currentNode = this.head
+
+    while(currentNode.next){
+      currentNode = currentNode.next
+    }
+      ++this._length
```
Comment (jrobcodes) [inline src/linkedList.js:27]: Indentation is confusing and makes it really hard to parse this code.

File: src/linkedList.js [hunk 49]
```diff
@@ -0,0 +1,112 @@
...
+  getTailNode(){ return this.tail }
+
+  contains(data) {
+    let currentNode = this.head
+
+    while(currentNode.next) {
+      if(currentNode.data === data) return true
+      currentNode = currentNode.next
+    }
+    if(this.getTailNode().data === data) return true
```
Comment (jrobcodes) [inline src/linkedList.js:45]: Here's another case of `if( true ) return true` - you can just return the result of the boolean expression directly:
```js
return this.getTailNode().getData() === data
```

File: src/linkedList.js [hunk 50]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  remove(data) {
+    let currentNode = this.head
+    while(currentNode.next.data === data) {
+      currentNode = currentNode.next
+    }
+    --this._length
+    return currentNode.next = this.tail = null
+    // delete this.tail
```
Comment (jrobcodes) [inline src/linkedList.js:95]: Remove comments

File: src/linkedList.js [hunk 51]
```diff
@@ -0,0 +1,112 @@
...
+  }
+
+  removesFirst() {
+    let currentNode = this.head
+    this.head = currentNode.next
+    --this._length
+  }
+
+  isEmpty() {
+    return this._length === 0 ? true : false
```
Comment (jrobcodes) [inline src/linkedList.js:105]:
```js
return this._length === 0
```

File: src/queue.js [hunk 52]
```diff
@@ -0,0 +1,34 @@
...
+    this.sequence = []
+  }
+
+  enqueue( data ) {
+    this.sequence[ this.head ] = data
+    this.head++
+  }
+
+  dequeue() {
+    return this.head > 0 ? this.sequence.splice(0, 1)[0] : null
```
Comment (jrobcodes) [inline src/queue.js:15]: Great use of builtins on array to implement this.  Consider rewriting to use `Node` objects, in order to better understand the `Queue` operations.

File: src/se.js [hunk 53]
```diff
@@ -0,0 +1,142 @@
...
+
+  isEmpty() {
+    return this.index === 0
+  }
+
+  contains( data ) {
+    for( let i = 0; i < this.index; i++ ){
+      if(this.sequence[i] === data){
+        return true
+      // }else if( this.sequence[i + 1] === undefined ){
```
Comment (jrobcodes) [inline src/se.js:21]: Remove comments

File: src/stack.js [hunk 54]
```diff
@@ -0,0 +1,24 @@
...
+    this.top = 0
+    this.element = []
+  }
+
+  push( data ) {
+    this.element[ this.top++ ] = data
+  }
+
+  pop() {
+    return this.top > 0 ? this.element.splice(-1, 1)[0] : null
```
Comment (jrobcodes) [inline src/stack.js:12]: I prefer the use if `if/else` over ternary operators, for readability.

**Review by jrobcodes — COMMENTED** [review]
Your testing code really needs more whitespace - code difficult for a human to parse when it is just a wall of text.  I suggest adding space between set up (creating the necessary objects for test), body (setting up those objects), and conclusion (the expectation/assertion).

When we pulled the repo, we encountered a number of typos that prevented us from being able to run the tests.  Once typos were fixed, 45 passed and 19 failed, the failures mostly in `Set` and `DoubleLinkedList`, both of which were marked as complete in the README.

I did not perform an in depth review of the graph implementation, as it was not marked as complete, and had some implementation that looked as if it relied on side effects. Happy to discuss in person.

Much of your code is lacking in whitespace - I realized as I was writing this comment that I had already written above.  Just wanted to leave it here to reinforce the need for whitespace.
